modules/plugins/completion/nvim-cmp/config.nix

```diff
@@ -3,91 +3,122 @@
   config,
   ...
 }: let
-  inherit (lib.modules) mkIf;
+  inherit (lib.modules) mkIf mkMerge;
   inherit (lib.strings) optionalString;
   inherit (lib.generators) mkLuaInline;
-  inherit (lib.nvim.dag) entryAfter;
   inherit (lib.nvim.lua) toLuaObject;
-  inherit (builtins) attrNames;
+  inherit (lib.nvim.attrsets) mapListToAttrs;
+  inherit (builtins) attrNames typeOf tryEval concatStringsSep;
 
   cfg = config.vim.autocomplete.nvim-cmp;
   luasnipEnable = config.vim.snippets.luasnip.enable;
+  getPluginName = plugin:
+    if typeOf plugin == "string"
+    then plugin
+    else if (plugin ? pname && (tryEval plugin.pname).success)
+    then plugin.pname
+    else plugin.name;
   inherit (cfg) mappings;
 in {
   config = mkIf cfg.enable {
     vim = {
-      startPlugins = [
-        "nvim-cmp"
-        "cmp-buffer"
-        "cmp-path"
-      ];
+      startPlugins = ["rtp-nvim"];
+      lazy.plugins = mkMerge [
+        (mapListToAttrs (package: {
+            name = getPluginName package;
+            value = {
+              inherit package;
+              lazy = true;
+              after = ''
+                local path = vim.fn.globpath(vim.o.packpath, 'pack/*/opt/${getPluginName package}')
+                require("rtp_nvim").source_after_plugin_dir(path)
```

Comment:
Why is this needed? Why can't you simply add the source plugins to startPlugins or optPlugins?

Reply:
cmp sources register themselves using an after/plugin/*.lua file

putting them in start plugin would make them also load cmp at startup

putting them in opt, lz.n won't source after/plugin files cuz packadd doesn't, so the sources aren't registered

Reply:
yeah no just abolish cmp at that point